add sync functionality to pelican database

gknapp1 opened the pull request:
Adds ability to sync data to Pelican database by extracting time series data:

- added functions in `File` class to extract time series data instead of spatial data
- add functionality in Pelican class
- added example for using `melt_pool_geometry_part` with Pelican data (currently, the only simulation type that supports Pelican sync)
- update thesis path generator to consider the wait time at the beginning of the scan track to appropriately register in time with time-based segmentation builds

Closes #80

gknapp1 commented on examples/melt_pool_geometry_part_pelican/template/Beam.txt:
We only need this template because of the differences between PBF and DED. Those differences caused the template that is already in the `thesis/melt_pool_geometry_part_pelican` app to not work properly. Still, definitely still something to address in future work.

gknapp1 commented on src/myna/core/files/file_melt_pool_geometry.py:
```diff
@@ -105,3 +106,37 @@ def get_values_for_sync(self, prefix="myna"):
+        value_units = ["m", "m", "m"]
+        return value_names, value_units
+
+    def get_timeseries_for_sync(
```
They have the same structure, but the fields are different. While this particular file type has both, I think that most file types will have one or the other.

streeve replied:
I was unclear - it seems like these two functions should call a generalized `get_fields_for_sync` where you just pass the list of strings and append into your final list

streeve replied:
Less generalizable in the future, but that would work too

streeve commented on src/myna/database/pelican.py:
```diff
@@ -403,3 +406,153 @@ def convert_dataframe_to_myna_scanpath(self, df, power=1):
+                    )
+                    sync_dict["time_segments"][metadata_key][
+                        "synced_on"
+                    ] = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
```
Could be some very general things to use from the base class instead

gknapp1 replied:
Moved sync metadata dictionary generation to base Database class.

streeve commented on src/myna/application/thesis/path.py:
```diff
@@ -84,29 +84,40 @@ def loadData(
-                    self.data.at[index, "time"] = 0.0
-                else:
-                    if self.data.at[index, "Mode"] == 1:
+                if self.data.at[index, "Mode"] == 1:
```
Describe these thesis changes in the description?

gknapp1 replied:
Added.

streeve reviewed (commented):
Am I better or worse than AI?

gknapp1 commented on src/myna/core/files/file_reduced_solidification.py:
```diff
@@ -53,15 +53,15 @@ def file_is_valid(self):
             return self.columns_are_valid(cols, expected_cols, expected_cols_types)
 
-    def get_values_for_sync(self, prefix="myna"):
+    def get_values_for_sync(self, mode="spatial"):
```
Updated comment and added error handling for =="temporal"

gknapp1 commented on src/myna/core/files/file_melt_pool_geometry.py:
```diff
@@ -54,54 +55,75 @@ def file_is_valid(self):
             return self.columns_are_valid(cols, expected_cols, expected_cols_types)
 
-    def get_names_for_sync(self, prefix="myna"):
+    def get_names_for_sync(self, mode="transient"):
```
I just set transient as the default, since it is better aligned with the data structure, but it could be either.

gknapp1 commented on src/myna/core/files/file_id.py:
```diff
@@ -46,15 +46,15 @@ def file_is_valid(self):
             return self.columns_are_valid(cols, expected_cols, expected_cols_types)
 
-    def get_values_for_sync(self, prefix="myna"):
+    def get_values_for_sync(self, mode="spatial"):
```
Updated comment and added error handling for =="temporal"

gknapp1 commented:
@streeve a moderately sized change, but I think the latest commit addresses your comments about the file sync functions. The `File` subclasses are much cleaner now--only the properties need to be defined for each subclass and all the functions are defined in the base class.

gknapp1 commented on src/myna/core/files/file.py:
```diff
@@ -43,10 +66,84 @@ def columns_are_valid(self, cols, expected_cols, expected_cols_types):
+            return locator, values, value_names, value_units
+
+        # If the filetype is not implemented, raise a NotImplementedError
         raise NotImplementedError
```
That's the case for now. We may want to have some sort of sync functionality for other types of files, e.g., VTK files, but I don't know what that should look like at the moment.

gknapp1 commented on src/myna/database/pelican.py:
```diff
@@ -403,3 +406,128 @@ def convert_dataframe_to_myna_scanpath(self, df, power=1):
+                    value_names = [f"{type_prefix}_{name}" for name in value_names]
+                    t_start = np.min(times)
+                    t_end = np.max(times)
+                except NotImplementedError:
```
This is currently the way for handling if the file has valid sync capability or not, so the try/except is for continuing even if the file doesn't have the ability to sync.
